File: android/src/main/java/com/giphyreactnativesdk/RNGiphyTheme.kt
package com.giphyreactnativesdk


import android.content.Context
import com.facebook.react.bridge.ReadableMap
import com.giphy.sdk.ui.themes.GPHCustomTheme
import com.giphy.sdk.ui.themes.GPHTheme
import com.giphy.sdk.ui.themes.Theme
import com.giphyreactnativesdk.utils.CaseConverter

val DEFAULT_PRESET = GPHTheme.Light


class RNGiphyTheme(context: Context, private var config: ReadableMap?) {
  private var preset: Theme

  init {
    val presetTheme = gphThemeByName(config?.getString("preset")) ?: DEFAULT_PRESET
    preset = presetTheme.getThemeResources(context)
  }

  fun applyToGPHCustomTheme() {
    GPHCustomTheme.apply {
      // region dialog's handle

      handleBarColor = getIntOrPreset("handleBarColor")

      // endregion

      // region emoji drawer

      emojiDrawerGradientTopColor = getIntOrPreset("emojiDrawerGradientTopColor")
      emojiDrawerGradientBottomColor = getIntOrPreset("emojiDrawerGradientBottomColor")
      emojiDrawerSeparatorColor = getIntOrPreset("emojiDrawerSeparatorColor")


      // endregion

      // region search bar

      searchBackButtonColor = getIntOrPreset("searchBackButtonColor")
      searchBarBackgroundColor = getIntOrPreset("searchBarBackgroundColor")
      searchTextColor = getIntOrPreset("searchTextColor")
      searchPlaceholderTextColor = getIntOrPreset("searchPlaceholderTextColor")

      // endregion

      // region suggestions

      suggestionCellBackgroundColor = getIntOrPreset("suggestionCellBackgroundColor")
      suggestionCellTextColor = getIntOrPreset("suggestionCellTextColor")

      // endregion

      // region tab bar

      tabBarSwitchDefaultColor = getIntOrPreset("tabBarSwitchDefaultColor")
      tabBarSwitchSelectedColor = getIntOrPreset("tabBarSwitchSelectedColor")

      // endregion

      // region confirmation

      confirmationSelectButtonColor = getIntOrPreset("confirmationSelectButtonColor")
      confirmationSelectButtonTextColor = getIntOrPreset("confirmationSelectButtonTextColor")
      confirmationBackButtonColor = getIntOrPreset("confirmationBackButtonColor")
      confirmationViewOnGiphyColor = getIntOrPreset("confirmationViewOnGiphyColor")

      // endregion

      // region other

      defaultTextColor = getIntOrPreset("defaultTextColor")
      dialogOverlayBackgroundColor = getIntOrPreset("dialogOverlayBackgroundColor")
      backgroundColor = getIntOrPreset("backgroundColor")
      usernameColor = getIntOrPreset("usernameColor")
      retryButtonBackgroundColor = getIntOrOptionalPreset("retryButtonBackgroundColor")
      retryButtonTextColor = getIntOrOptionalPreset("retryButtonTextColor")

      // endregion
    }
  }

  private fun getInt(field: String): Int? {
    if (config == null || config?.hasKey(field) == false) {
      return null
    }
    return config!!.getInt(field)
  }

  private fun getIntOrPreset(field: String): Int {
    return getInt(field) ?: getPresetField(field) as Int
  }

  private fun getIntOrOptionalPreset(field: String): Int? {
    return getInt(field) ?: getPresetField(field) as Int?
  }

  private fun getPresetField(field: String): Any? {
    return when (field) {
      "handleBarColor" -> preset.handleBarColor
      "emojiDrawerGradientTopColor" -> preset.emojiDrawerGradientTopColor
      "emojiDrawerGradientBottomColor" -> preset.emojiDrawerGradientBottomColor
      "emojiDrawerSeparatorColor" -> preset.emojiDrawerSeparatorColor
      "searchBarBackgroundColor" -> preset.searchBarBackgroundColor
      "searchTextColor" -> preset.searchTextColor
      "searchButtonIcon" -> preset.searchButtonIcon
      "searchBackButtonColor" -> preset.searchBackButtonColor
      "searchPlaceholderTextColor" -> preset.searchPlaceholderTextColor
      "suggestionCellBackgroundColor" -> preset.suggestionCellBackgroundColor
      "suggestionCellTextColor" -> preset.suggestionCellTextColor
      "tabBarSwitchDefaultColor" -> preset.tabBarSwitchDefaultColor
      "tabBarSwitchSelectedColor" -> preset.tabBarSwitchSelectedColor
      "confirmationSelectButtonColor" -> preset.confirmationSelectButtonColor
      "confirmationSelectButtonTextColor" -> preset.confirmationSelectButtonTextColor
      "confirmationBackButtonColor" -> preset.confirmationBackButtonColor
      "confirmationViewOnGiphyColor" -> preset.confirmationViewOnGiphyColor
      "backgroundColor" -> preset.backgroundColor
      "dialogOverlayBackgroundColor" -> preset.dialogOverlayBackgroundColor
      "defaultTextColor" -> preset.defaultTextColor
      "usernameColor" -> preset.usernameColor
      "retryButtonBackgroundColor" -> preset.retryButtonBackgroundColor
      "retryButtonTextColor" -> preset.retryButtonTextColor
      else -> {
        throw IllegalArgumentException("Unknown field: $field")
      }
    }
  }

  companion object {
    private fun gphThemeByName(name: String?): GPHTheme? {
      if (name == null) {
        return null
      }
      return GPHTheme.values().firstOrNull {
        it.name == CaseConverter.capitalize(name)
      }
    }
  }
}

File: ios/RNGiphyTheme.swift
import UIKit

import GiphyUISDK

let THEME_COLOR_FIELDS = [
  "avatarPlaceholderColor",
  "backgroundColor",
  "backgroundColorForLoadingCells",
  "confirmationBackButtonColor",
  "confirmationSelectButtonColor",
  "confirmationSelectButtonTextColor",
  "confirmationViewOnGiphyColor",
  "defaultTextColor",
  "dialogOverlayBackgroundColor",
  "emojiDrawerGradientBottomColor",
  "emojiDrawerGradientTopColor",
  "emojiDrawerSeparatorColor",
  "handleBarColor",
  "searchBackButtonColor",
  "searchBarBackgroundColor",
  "searchPlaceholderTextColor",
  "searchTextColor",
  "stickerBackgroundColor",
  "suggestionCellBackgroundColor",
  "suggestionCellTextColor",
  "tabBarSwitchDefaultColor",
  "tabBarSwitchSelectedColor",
  "usernameColor",
]

private func deserializeRNConfig(_ rawConfig: NSDictionary?) -> NSDictionary {
  let config = NSMutableDictionary(dictionary: rawConfig ?? [:])

  if let preset = config["preset"] as? String {
    config["preset"] = GPHThemeType.fromRNValue(preset)
  }
  if let kbAppearance = config["keyboardAppearance"] as? String {
    config["keyboardAppearance"] = UIKeyboardAppearance.fromRNValue(kbAppearance)
  }
  if let edScrollIndicatorStyle = config["emojiDrawerScrollIndicatorStyle"] as? String {
    config["emojiDrawerScrollIndicatorStyle"] = UIScrollView.IndicatorStyle.fromRNValue(edScrollIndicatorStyle)
  }

  THEME_COLOR_FIELDS.forEach { field in
    if config[field] != nil {
      config[field] = RCTConvert.uiColor(config[field])
    }
  }

  return config
}

public class RNGiphyTheme: GPHTheme {
  var config: NSDictionary

  public init(rnConfig: NSDictionary?) {
    config = deserializeRNConfig(rnConfig)
    super.init()
    type = config["preset"] as? GPHThemeType ?? .dark
  }

  // MARK: - Dialog's Handle
  public override var handleBarColor: UIColor {
    config["handleBarColor"] as? UIColor ?? super.handleBarColor
  }

  // MARK: - Emoji Drawer
  public override var emojiDrawerScrollIndicatorStyle: UIScrollView.IndicatorStyle {
    config["emojiDrawerScrollIndicatorStyle"] as? UIScrollView.IndicatorStyle ?? super.emojiDrawerScrollIndicatorStyle
  }

  public override var emojiDrawerGradientTopColor: UIColor {
    config["emojiDrawerGradientTopColor"] as? UIColor ?? super.emojiDrawerGradientTopColor
  }

  public override var emojiDrawerGradientBottomColor: UIColor {
    config["emojiDrawerGradientBottomColor"] as? UIColor ?? super.emojiDrawerGradientBottomColor
  }

  public override var emojiDrawerSeparatorColor: UIColor {
    config["emojiDrawerSeparatorColor"] as? UIColor ?? super.emojiDrawerSeparatorColor
  }

  // MARK: - Search Bar
  public override var searchBackButtonColor: UIColor {
    config["searchBackButtonColor"] as? UIColor ?? super.searchBackButtonColor
  }

  public override var searchPlaceholderTextColor: UIColor {
    config["searchPlaceholderTextColor"] as? UIColor ?? super.searchPlaceholderTextColor
  }

  public override var searchTextColor: UIColor {
    config["searchTextColor"] as? UIColor ?? super.searchTextColor
  }

  public override var searchBarBackgroundColor: UIColor {
    config["searchBarBackgroundColor"] as? UIColor ?? super.searchBarBackgroundColor
  }

  public override var searchBarPadding: CGFloat {
    config["searchBarPadding"] as? CGFloat ?? super.searchBarPadding
  }

  public override var searchBarCornerRadius: CGFloat {
    config["searchBarCornerRadius"] as? CGFloat ?? super.searchBarCornerRadius
  }

  // MARK: - Suggestions
  public override var showSuggestionsBar: Bool {
    config["showSuggestionsBar"] as? Bool ?? super.showSuggestionsBar
  }

  public override var suggestionCellBackgroundColor: UIColor {
    config["suggestionCellBackgroundColor"] as? UIColor ?? super.suggestionCellBackgroundColor
  }

  public override var suggestionCellTextColor: UIColor {
    config["suggestionCellTextColor"] as? UIColor ?? super.suggestionCellTextColor
  }

  // MARK: - Tab Bar
  public override var tabBarSwitchSelectedColor: UIColor {
    config["tabBarSwitchSelectedColor"] as? UIColor ?? super.tabBarSwitchSelectedColor
  }

  public override var tabBarSwitchDefaultColor: UIColor? {
    config["tabBarSwitchDefaultColor"] as? UIColor ?? super.tabBarSwitchDefaultColor
  }

  public override var tabBarBackgroundAlpha: CGFloat {
    config["tabBarBackgroundAlpha"] as? CGFloat ?? super.tabBarBackgroundAlpha
  }

  // MARK: - Confirmation
  public override var confirmationSelectButtonColor: UIColor {
    config["confirmationSelectButtonColor"] as? UIColor ?? super.confirmationSelectButtonColor
  }

  public override var confirmationSelectButtonTextColor: UIColor {
    config["confirmationSelectButtonTextColor"] as? UIColor ?? super.confirmationSelectButtonTextColor
  }

  public override var usernameColor: UIColor {
    config["usernameColor"] as? UIColor ?? super.usernameColor
  }

  public override var confirmationBackButtonColor: UIColor {
    config["confirmationBackButtonColor"] as? UIColor ?? super.confirmationBackButtonColor
  }

  public override var confirmationViewOnGiphyColor: UIColor {
    config["confirmationViewOnGiphyColor"] as? UIColor ?? super.confirmationViewOnGiphyColor
  }

  public override var avatarPlaceholderColor: UIColor {
    config["avatarPlaceholderColor"] as? UIColor ?? super.avatarPlaceholderColor
  }

  // MARK: - Grid Content
  public override var cellCornerRadius: CGFloat {
    config["cellCornerRadius"] as? CGFloat ?? super.cellCornerRadius
  }

  public override var stickerBackgroundColor: UIColor {
    config["stickerBackgroundColor"] as? UIColor ?? super.stickerBackgroundColor
  }

  public override var fixedSizeCells: Bool {
    config["fixedSizeCells"] as? Bool ?? super.fixedSizeCells
  }

  public override var backgroundColorForLoadingCells: UIColor {
    config["backgroundColorForLoadingCells"] as? UIColor ?? super.backgroundColorForLoadingCells
  }

  // MARK: - Keyboard
  public override var keyboardAppearance: UIKeyboardAppearance {
    config["keyboardAppearance"] as? UIKeyboardAppearance ?? super.keyboardAppearance
  }

  // MARK: - Other
  public override var backgroundColor: UIColor {
    config["backgroundColor"] as? UIColor ?? super.backgroundColor
  }
}
